File: scripts/check-duplicates-against-database.py
#!/usr/bin/env python3
"""
Compare CSV file with live Supabase database to find and remove duplicates
"""

import csv
import sys
import os
import re
from typing import List, Dict, Set, Tuple
from dotenv import load_dotenv

# Load environment variables
load_dotenv('.env.local')

try:
    from supabase import Client, create_client
except ImportError:
    print("Error: supabase-py not installed. Install with: pip install supabase")
    sys.exit(1)

def normalize_property_name(name: str) -> str:
    """Normalize property name for comparison."""
    if not name:
        return ""
    
    # Normalize to lowercase and strip whitespace
    normalized = name.lower().strip()
    
    # Remove common variations: dashes, parentheses with state codes, extra spaces
    normalized = normalized.replace('-', ' ')
    # Remove anything in parentheses
    normalized = re.sub(r'\([^)]*\)', '', normalized)
    # Normalize multiple spaces to single space
    normalized = ' '.join(normalized.split())
    
    return normalized


def get_database_properties() -> Set[str]:
    """Query Supabase database for all unique property names."""
    supabase_url = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
    # Try both naming conventions for the service role key
    secret_key = os.getenv('SUPABASE_SERVICE_ROLE_KEY') or os.getenv('SUPABASE_SECRET_KEY')
    
    if not supabase_url or not secret_key:
        print('Error: Supabase credentials not found in .env.local')
        print('Required: NEXT_PUBLIC_SUPABASE_URL and either SUPABASE_SERVICE_ROLE_KEY or SUPABASE_SECRET_KEY')
        sys.exit(1)
    
    try:
        supabase: Client = create_client(supabase_url, secret_key)
        
        print('📥 Fetching all property names from database...')
        
        # Fetch all unique property names (in batches)
        all_properties = []
        offset = 0
        batch_size = 1000
        has_more = True
        
        while has_more:
            response = supabase.table('all_glamping_properties')\
                .select('property_name')\
                .not_('property_name', 'is', None)\
                .range(offset, offset + batch_size - 1)\
                .execute()
            
            if not response.data:
                break
            
            # Extract property names
            property_names = [
                row.get('property_name', '').strip() 
                for row in response.data 
                if row.get('property_name', '').strip()
            ]
            
            all_properties.extend(property_names)
            offset += batch_size
            has_more = len(response.data) == batch_size
            
            print(f'  Fetched {len(all_properties)} property names...')
        
        # Get unique property names
        unique_properties = set(all_properties)
        print(f'✅ Found {len(unique_properties)} unique property names in database\n')
        
        return unique_properties
        
    except Exception as e:
        print(f'Error querying database: {e}')
        sys.exit(1)


def find_match_in_database(
    csv_property_name: str,
    db_properties: Set[str]
) -> Tuple[bool, str]:
    """Check if a CSV property matches any database property."""
    normalized_csv = normalize_property_name(csv_property_name)
    
    # Create normalized mapping
    normalized_db_map = {normalize_property_name(p): p for p in db_properties}
    
    # Direct exact match
    if normalized_csv in normalized_db_map:
        return True, normalized_db_map[normalized_csv]
    
    # Extract the location/region name (everything after brand name)
    brand_patterns = [
        'postcard cabins',
        'huttopia',
        'under canvas',
        'glamping.com',
        'field mag',
        'us news travel'
    ]
    
    csv_location = normalized_csv
    for brand in brand_patterns:
        if normalized_csv.startswith(brand):
            csv_location = normalized_csv[len(brand):].strip()
            break
    
    # Check if any database property has the same location
    for db_normalized, db_original in normalized_db_map.items():
        db_location = db_normalized
        for brand in brand_patterns:
            if db_normalized.startswith(brand):
                db_location = db_normalized[len(brand):].strip()
                break
        
        # Exact location match
        if csv_location and db_location and csv_location == db_location:
            return True, db_original
        
        # Check if locations match (exact or one contains the other)
        if csv_location and db_location and len(csv_location) > 3 and len(db_location) > 3:
            # Exact match after cleaning
            if csv_location == db_location:
                return True, db_original
            # One contains the other (for cases like "Shenandoah" vs "Shenandoah North")
            if csv_location in db_location or db_location in csv_location:
                # Make sure it's not too short to avoid false matches
                if len(csv_location) > 5 and len(db_location) > 5:
                    return True, db_original
    
    return False, ""


def main():
    csv_file = 'csv/glamping-com-north-america-missing-properties.csv'
    
    print('=' * 70)
    print('Comparing CSV with Live Database to Find Duplicates')
    print('=' * 70)
    print()
    
    # Read CSV file
    print(f'📖 Reading CSV file: {csv_file}')
    csv_properties = []
    fieldnames = None
    try:
        with open(csv_file, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            fieldnames = list(reader.fieldnames)
            csv_properties = list(reader)
    except Exception as e:
        print(f'Error reading CSV: {e}')
        sys.exit(1)
    
    print(f'✅ Found {len(csv_properties)} properties in CSV\n')
    
    # Get database properties
    db_properties = get_database_properties()
    
    # Check each CSV property against database
    print('🔍 Checking for duplicates...\n')
    duplicates: List[Dict] = []
    unique: List[Dict] = []
    
    for row in csv_properties:
        property_name = row.get('Property Name', '').strip()
        if not property_name:
            unique.append(row)
            continue
        
        found, match = find_match_in_database(property_name, db_properties)
        
        if found:
            print(f'  ✗ DUPLICATE: {property_name}')
            print(f'    Matches database: {match}')
            duplicates.append({'row': row, 'match': match})
        else:
            unique.append(row)
    
    print()
    print('📊 Summary:')
    print(f'   Total in CSV: {len(csv_properties)}')
    print(f'   Duplicates found: {len(duplicates)}')
    print(f'   Unique properties: {len(unique)}')
    print()
    
    # Remove duplicates from CSV if any found
    if len(duplicates) > 0:
        print('🗑️  Removing duplicates from CSV...\n')
        
        # Write updated CSV
        with open(csv_file, 'w', encoding='utf-8', newline='') as f:
            writer = csv.DictWriter(f, fieldnames=fieldnames)
            writer.writeheader()
            writer.writerows(unique)
        
        print(f'✅ Updated CSV file: {csv_file}')
        print(f'   Removed {len(duplicates)} duplicate properties')
        print(f'   Remaining properties: {len(unique)}')
        print()
        print('Removed duplicates:')
        for i, dup in enumerate(duplicates, 1):
            print(f'  {i}. {dup["row"]["Property Name"]} (matched: {dup["match"]})')
    else:
        print('✅ No duplicates found - CSV is clean!')
    
    print()
    print('=' * 70)
    print('Analysis complete!')
    print('=' * 70)


if __name__ == '__main__':
    main()

File: scripts/check-duplicates-against-database.ts
#!/usr/bin/env ts-node
/**
 * Compare CSV file with live Supabase database to find and remove duplicates
 */

import { config } from 'dotenv';
import { resolve } from 'path';
import { createClient } from '@supabase/supabase-js';
import * as fs from 'fs';
import * as csv from 'csv-parse/sync';
import { stringify } from 'csv-stringify/sync';

// Load environment variables from .env.local
config({ path: resolve(process.cwd(), '.env.local') });

interface CSVRow {
  [key: string]: string;
}

async function getDatabaseProperties() {
  const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
  const secretKey = process.env.SUPABASE_SECRET_KEY;

  if (!supabaseUrl || !secretKey) {
    throw new Error('Supabase credentials not found in environment variables');
  }

  const supabase = createClient(supabaseUrl, secretKey, {
    auth: {
      persistSession: false,
      autoRefreshToken: false,
    },
  });

  console.log('📥 Fetching all property names from database...');

  // Fetch all unique property names from database
  let allProperties: string[] = [];
  let offset = 0;
  const batchSize = 1000;
  let hasMore = true;

  while (hasMore) {
    const { data, error } = await supabase
      .from('all_glamping_properties')
      .select('property_name')
      .not('property_name', 'is', null)
      .range(offset, offset + batchSize - 1);

    if (error) {
      throw new Error(`Error fetching from database: ${error.message}`);
    }

    if (!data || data.length === 0) {
      break;
    }

    // Extract unique property names
    const propertyNames = data
      .map((row: any) => row.property_name?.trim())
      .filter((name: string) => name && name.length > 0);

    allProperties = allProperties.concat(propertyNames);
    offset += batchSize;
    hasMore = data.length === batchSize;

    console.log(`  Fetched ${allProperties.length} property names...`);
  }

  // Get unique property names
  const uniqueProperties = new Set(allProperties);
  console.log(`✅ Found ${uniqueProperties.size} unique property names in database\n`);

  return uniqueProperties;
}

function normalizePropertyName(name: string): string {
  if (!name) return '';
  
  // Normalize to lowercase and strip whitespace
  let normalized = name.toLowerCase().trim();
  
  // Remove common variations: dashes, parentheses with state codes, extra spaces
  normalized = normalized.replace(/-/g, ' ');
  normalized = normalized.replace(/\([^)]*\)/g, ''); // Remove anything in parentheses
  normalized = normalized.replace(/\s+/g, ' '); // Normalize multiple spaces
  
  return normalized.trim();
}

function findMatchInDatabase(
  csvPropertyName: string,
  dbProperties: Set<string>
): { found: boolean; match?: string } {
  const normalizedCsv = normalizePropertyName(csvPropertyName);

  // Check for exact match
  for (const dbProp of dbProperties) {
    const normalizedDb = normalizePropertyName(dbProp);
    
    if (normalizedCsv === normalizedDb) {
      return { found: true, match: dbProp };
    }
  }

  // Check for partial matches (location name matches)
  const csvLocation = normalizedCsv.replace(/^(postcard cabins|huttopia|under canvas|glamping\.com|field mag|us news travel)\s*/i, '').trim();
  
  if (csvLocation.length > 3) {
    for (const dbProp of dbProperties) {
      const normalizedDb = normalizePropertyName(dbProp);
      const dbLocation = normalizedDb.replace(/^(postcard cabins|huttopia|under canvas|glamping\.com|field mag|us news travel)\s*/i, '').trim();
      
      // Check if locations match (exact or one contains the other)
      if (dbLocation.length > 3) {
        if (csvLocation === dbLocation) {
          return { found: true, match: dbProp };
        }
        // Check if one contains the other (for variations like "Shenandoah" vs "Shenandoah North")
        if (csvLocation.length > 5 && dbLocation.length > 5) {
          if (csvLocation.includes(dbLocation) || dbLocation.includes(csvLocation)) {
            return { found: true, match: dbProp };
          }
        }
      }
    }
  }

  return { found: false };
}

async function main() {
  const csvFile = 'csv/glamping-com-north-america-missing-properties.csv';

  console.log('='.repeat(70));
  console.log('Comparing CSV with Live Database to Find Duplicates');
  console.log('='.repeat(70));
  console.log();

  // Read CSV file
  console.log(`📖 Reading CSV file: ${csvFile}`);
  const csvContent = fs.readFileSync(csvFile, 'utf-8');
  const csvRows: CSVRow[] = csv.parse(csvContent, {
    columns: true,
    skip_empty_lines: true,
  });

  console.log(`✅ Found ${csvRows.length} properties in CSV\n`);

  // Get database properties
  const dbProperties = await getDatabaseProperties();

  // Check each CSV property against database
  console.log('🔍 Checking for duplicates...\n');
  const duplicates: Array<{ row: CSVRow; match: string }> = [];
  const unique: CSVRow[] = [];

  for (const row of csvRows) {
    const propertyName = row['Property Name']?.trim();
    if (!propertyName) {
      continue;
    }

    const matchResult = findMatchInDatabase(propertyName, dbProperties);

    if (matchResult.found) {
      console.log(`  ✗ DUPLICATE: ${propertyName}`);
      console.log(`    Matches database: ${matchResult.match}`);
      duplicates.push({ row, match: matchResult.match! });
    } else {
      unique.push(row);
    }
  }

  console.log();
  console.log(`📊 Summary:`);
  console.log(`   Total in CSV: ${csvRows.length}`);
  console.log(`   Duplicates found: ${duplicates.length}`);
  console.log(`   Unique properties: ${unique.length}`);
  console.log();

  // Remove duplicates from CSV if any found
  if (duplicates.length > 0) {
    console.log('🗑️  Removing duplicates from CSV...\n');
    
    // Get fieldnames from first row
    const fieldnames = Object.keys(csvRows[0]);

    // Write updated CSV
    const updatedCsv = stringify(unique, {
      header: true,
      columns: fieldnames,
    });

    fs.writeFileSync(csvFile, updatedCsv, 'utf-8');

    console.log(`✅ Updated CSV file: ${csvFile}`);
    console.log(`   Removed ${duplicates.length} duplicate properties`);
    console.log(`   Remaining properties: ${unique.length}`);
    console.log();
    console.log('Removed duplicates:');
    duplicates.forEach((dup, i) => {
      console.log(`  ${i + 1}. ${dup.row['Property Name']} (matched: ${dup.match})`);
    });
  } else {
    console.log('✅ No duplicates found - CSV is clean!');
  }

  console.log();
  console.log('='.repeat(70));
  console.log('Analysis complete!');
  console.log('='.repeat(70));
}

// Run the script
main().catch((error) => {
  console.error('❌ Error:', error);
  process.exit(1);
});
